schema: `metadata` at the top level

supakeen opened the pull request:
A new metadata key at the top level which contains a generators key which is a list of all generators involved in the creation of the manifest.

Tools that generate manifests are, for example: [otk](https://github.com/osbuild/otk), [osbuild-composer](https://github.com/osbuild/osbuild-composer) (through [images](https://github.com/osbuild/images)).

supakeen commented:
Now has a test; one of the test manifests includes the `meta` key at the top level.

ondrejbudai reviewed (commented):
I'm in favor of this idea. Thanks!

mvo5 reviewed (commented):
I guess this comes down to the question how precise/loose we want to be. The (obvious) advantage of a free-form meta is that the generators can store anything they want and osbuild does not need to know about keys etc. The downside is the same, we will not be able to even spot typos if this is totally free form.

It might be nice to have some examples before we decide. The obviuos one would be:
- generator, generator-version

What else would we put here?

achilleas-k commented:
> The obviuos one would be:
> 
> * generator, generator-version
> 
> What else would we put here?

```json
{
  "generator": {
    "name": "osbuild/images",
    "version": "v0.90",
    "timestamp": "1728905899"
  }
}
```

supakeen commented:
That'd be one example; another example could be:

- customizations used to generate the image
- image type
- pre-depsolve package set
- an original omnifest
- the sbom related to everything inside the image (though we might want to move that under a structured field)

supakeen commented:
@mvo5 my idea would be to have `meta` be there and not depend on it (since it doesn't have a  schema), start there with conventions to see if we like things and afterwards propose things for inclusion at the top level schema.

achilleas-k commented:
> Thanks for listing some of the planned content. I must say this makes me very uneasy. We have this very nice osbuild json format which is versioned and validated. Now we are adding a "meta" tree that can contain anything (and some of the example mean a lot of extra data).
> 
> I wonder if instead of putting this all into the (so far) well defined osbuild manifest we should create a new container format that includes this additional typed data (like the otk file) and in addition the osbuild manifest. This could be as simple as a tar file.
> 
> I like "meta" but I dislike that it's not typed, maybe we can find a compromise, add some meta keys (like generator.{name,version,timestamp}) into our schema and put the rest into this new container?

Thanks for this comment, it nicely expresses my feelings about the change.

I'm finding it hard to explain exactly what makes me uneasy about this.  My thinking keeps cycling through:
- This is fine because it doesn't affect the functionality or the identity of the manifest.
- If it doesn't affect the functionality or identify of the manifest, then what is it for and why does it need to be part of the manifest?
- It's metadata tied to a manifest that could be used by other tools to describe the origin or purpose of the manifest (or anything).
- If it's meant for other tools, maybe those tools should have their own manifest-adjacent, external metadata for its manifests.
- But keeping them together means the metadata is more tightly bound to the manifest itself.

> start there with conventions to see if we like things and afterwards propose things for inclusion at the top level schema.

Starting wide open and locking down never works, in my experience.  The opposite does though.

I think we should consider the purpose and start with at least some restrictions.  One thing that would make sense to me is to at least restrict the top level type to strings, and maybe even some of the structure of each element beneath that.  So, for example, I would enforce something like:
```json
{
  "meta": {
    "title": "Arbitrary meta information keyed by project identifier",
    "type": "object",
    "additionalProperties": false,
    "patternProperties": {
      "^.*$": {
        "type": "object",
        "additionalProperties": false,
        "properties": {
          "description": {
            "type": "string"
          },
          "metadata": {
            "type": "object",
            "additionalProperties": true
          }
        }
      }
    }
  }
}
```

(This is just a quick recommendation of what I'm considering and I don't expect it to be taken without discussion or modification)

My thinking here is that it enforces, at least, that top-level keys under `"meta"` are strings and suggests that each tool should add its name as the top-level key and a description below that, alongside the metadata it wants to store.  So we could have something like:
```json
{
  "meta": {
    "osbuild-composer": {
      "description": "Manifest created by osbuild-composer v101 (https://github.com/osbuild/osbuild-composer)",
      "metadata": {
        "version": "101",
        "created": "2024-10-17 14:05",
        "emoji of the day": "💥"
      }
    },
    "osbuild-depsolve-dnf": {
      "description": "Package versions and dependencies resolved using osbuild-depsolve-dnf from osbuild v120 (https://github.com/osbuild/osbuild)",
      "metadata": {
        "version": "120",
        "package selection": {
          "build": [ "..." ],
          "os": [ "..." ]
        }
      }
    }
  }
}
```

supakeen commented:
@mvo5 / @achilleas-k, since we/I only have a concrete example for the generator I propose the following:

```
metadata:
  generator:
    [strkey]:
      version: string
```

This would allow for multiple generators to save their information if that's something we want in the future. If we go directly with `name`/`version` under generator we'd preclude that.

Is that OK?

mvo5 commented:
> @mvo5 / @achilleas-k, since we/I only have a concrete example for the generator I propose the following:
> 
> ```
> metadata:  # instead of meta
>   generator:
>     [strkey]:  # so we can store information for multiple if that ever applies
>       version: string
> ```
> 
> This would allow for multiple generators to save their information if that's something we want in the future. If we go directly with `name`/`version` under generator we'd preclude that.
> 
> Is that OK?

That works for me, fwiw, I don't have an opinion on "meta" vs "metadata" (the later is a bit more specific which is probably nicer). I am also fine with version/timestamp(or created)/vcs_revision as potential fields as those seem to have obvious use-cases.

supakeen commented:
Updated and adjusted; there's now a new metadata key at the top level which contains a generators key which is a list of all generators involved in the creation of the manifest.

mvo5 commented on schemas/osbuild2.json:
```diff
@@ -83,6 +86,24 @@
+      "type": "object",
+      "additionalProperties": false,
+      "properties": {
+        "generator": {
```
(super nitpick) if it's an array maybe use plural here, i.e. `generators` ?

P.S. I'm not 100% sold on the multiple generators but I don't mind much either :)

mvo5 reviewed (commented):
This is very nice now, one tiny suggesiton inline and tests are failing(I have not looked in detail why though)

supakeen commented:
> This is very nice now, one tiny suggesiton inline and tests are failing(I have not looked in detail why though)

Because I made a booboo that I've fixed :)

mvo5 reviewed (commented):
Thanks! This looks fine now, Sorry that I drag this on further :( But I had one interesting observation while reviewing.

I added the following snippet to the existing format test:
```python
diff --git a/test/mod/test_fmt_v2.py b/test/mod/test_fmt_v2.py
index 65f7d3ba..f1014850 100644
--- a/test/mod/test_fmt_v2.py
+++ b/test/mod/test_fmt_v2.py
@@ -12,6 +12,14 @@ import osbuild.meta
 
 BASIC_PIPELINE = {
     "version": "2",
+    "metadata": {
+        "generators": [
+            {
+                "name": "handcrafted",
+                "version": "NaN",
+            }
+        ],
+    },
     "sources": {
         "org.osbuild.curl": {
             "items": {

```
and when doing this I saw that the `test_describe()` fails because the manifest is not reproduced with "describe()", the "metadata" object is omited - it seems osbuild/format/v2.py would need a new "def describe metadata" (or something like this).

It's an interesting question if `--inspect` should include the metadata or not (maybe @achilleas-k has an opinions here?). I personally have not used `--inspect` myself. My gut feeling is that we should reproduce the manifest as faithfully as possible and include the metadata but I can see an argument against this. (But I can see the other side too) If we not include metadata in describe(), it is probably good to add a comment in osbuild/format/v2.py about this, something like `# "metadata" object not included as it is not required for the functionality of osbuild`.

supakeen commented:
> Thanks! This looks fine now, Sorry that I drag this on further :( But I had one interesting observation while reviewing.
> 
> I added the following snippet to the existing format test:
> 
> ```python
> diff --git a/test/mod/test_fmt_v2.py b/test/mod/test_fmt_v2.py
> index 65f7d3ba..f1014850 100644
> --- a/test/mod/test_fmt_v2.py
> +++ b/test/mod/test_fmt_v2.py
> @@ -12,6 +12,14 @@ import osbuild.meta
>  
>  BASIC_PIPELINE = {
>      "version": "2",
> +    "metadata": {
> +        "generators": [
> +            {
> +                "name": "handcrafted",
> +                "version": "NaN",
> +            }
> +        ],
> +    },
>      "sources": {
>          "org.osbuild.curl": {
>              "items": {
> ```
> 
> and when doing this I saw that the `test_describe()` fails because the manifest is not reproduced with "describe()", the "metadata" object is omited - it seems osbuild/format/v2.py would need a new "def describe metadata" (or something like this).
> 
> It's an interesting question if `--inspect` should include the metadata or not (maybe @achilleas-k has an opinions here?). I personally have not used `--inspect` myself. My gut feeling is that we should reproduce the manifest as faithfully as possible and include the metadata but I can see an argument against this. (But I can see the other side too) If we not include metadata in describe(), it is probably good to add a comment in osbuild/format/v2.py about this, something like `# "metadata" object not included as it is not required for the functionality of osbuild`.

This is a really good find. It's better to start with exactly reproducing the manifest. I'll update this PR to include the change you made in `test_fmt_v2.py` and adjust the describe-parts.

mvo5 reviewed (approved):
Thank you

supakeen commented:
Ok, I've re-ordered the imports by hand (nothing else changed) to what they should be but the above is worrisome.

bcl reviewed (approved):
Looks good to me.
